pkg/controller/endpointslicemirroring/reconciler.go

```diff
@@ -0,0 +1,287 @@
+		if _, ok := d.endpointsByKey[portKey]; !ok {
+			for _, existingSlice := range existingSlices {
+				slices.toDelete = append(slices.toDelete, existingSlice)
+			}
+		}
+	}
+
+	metrics.EndpointsAddedPerSync.WithLabelValues().Observe(float64(totals.added))
+	metrics.EndpointsUpdatedPerSync.WithLabelValues().Observe(float64(totals.updated))
+	metrics.EndpointsRemovedPerSync.WithLabelValues().Observe(float64(totals.removed))
+
+	endpointsNN := types.NamespacedName{Name: endpoints.Name, Namespace: endpoints.Namespace}
+	r.metricsCache.UpdateEndpointPortCache(endpointsNN, epMetrics)
+
+	return r.finalize(endpoints, slices)
+}
+
+// reconcileByPortMapping compares the endpoints found in existing slices with
+// the list of desired endpoints and returns lists of slices to create, update,
+// and delete.
+func (r *reconciler) reconcileByPortMapping(
+	endpoints *corev1.Endpoints,
+	existingSlices []*discovery.EndpointSlice,
+	desiredSet endpointSet,
+	endpointPorts []discovery.EndpointPort,
+	addressType discovery.AddressType,
+) (slicesByAction, totalsByAction) {
+	slices := slicesByAction{}
+	totals := totalsByAction{}
+
+	// If no endpoints are desired, mark existing slices for deletion and
+	// return.
+	if desiredSet.Len() == 0 {
+		slices.toDelete = existingSlices
+		for _, epSlice := range existingSlices {
+			totals.removed += len(epSlice.Endpoints)
+		}
+		return slices, totals
+	}
+
+	if len(existingSlices) == 0 {
+		// if no existing slices, all desired endpoints will be added.
+		totals.added = desiredSet.Len()
+	} else {
+		// if >0 existing slices, mark all but 1 for deletion.
+		slices.toDelete = existingSlices[1:]
+
+		// Return early if first slice matches desired endpoints.
+		totals = totalChanges(existingSlices[0], desiredSet)
+		if totals.added == 0 && totals.updated == 0 && totals.removed == 0 {
+			return slices, totals
+		}
+	}
+
+	// generate a new slice with the desired endpoints.
+	newSlice := newEndpointSlice(endpoints, endpointPorts, addressType)
+	for desiredSet.Len() > 0 && len(newSlice.Endpoints) < int(r.maxEndpointsPerSubset) {
+		endpoint, _ := desiredSet.PopAny()
+		newSlice.Endpoints = append(newSlice.Endpoints, *endpoint)
+	}
+
+	if len(existingSlices) > 0 {
+		newSlice.Name = existingSlices[0].Name
+		slices.toUpdate = []*discovery.EndpointSlice{newSlice}
+	} else {
+		slices.toCreate = []*discovery.EndpointSlice{newSlice}
+	}
+	return slices, totals
+}
+
+// finalize creates, updates, and deletes slices as specified
+func (r *reconciler) finalize(endpoints *corev1.Endpoints, slices slicesByAction) error {
+	errs := []error{}
+
+	// If there are slices to create and delete, recycle the slices marked for
+	// deletion by replacing creates with updates of slices that would otherwise
+	// be deleted.
+	recycleSlices(&slices)
+
+	// Don't create more EndpointSlices if corresponding Endpoints resource is
+	// being deleted.
+	if endpoints.DeletionTimestamp == nil {
```

Comment:
The semantics of this is little weird.

Imagine that the corresponding Endpoint has been marked for deletion and the user scales up their endpoints after that. Those endpoints will randomly be part of the resultant set. We are still updating the endpoint set, but not in its entirety. 

Does it make more sense to not have special behavior for non-Finalized resources? This makes it harder reason about. Are we being too clever here?

Reply:
This is mirroring the approach we decided on in https://github.com/kubernetes/kubernetes/pull/91311. The goal here is to not conflict with garbage collection while still supporting existing EndpointSlices in the case deletion takes a long time. In the case of this controller with a 1:1 mapping between EndpointSlice and EndpointSubset, it seems unlikely that an EndpointSlice would be removed for any reason other than garbage collection when there is a non nil deletion timestamp.